[DPE-5555] Security cryptography documentation (#240)
canonical/kafka-operator: docs/explanation/e-cryptography.md

# Cryptography

> 

Reviewer: For example, we can use an introduction:
```suggestion

This document describes cryptography used by Charmed Kafka.

```

## Resource checksums

Every version of the Charmed Kafka and Charmed ZooKeeper operators install a pinned revision of the Charmed Kafka snap
and Charmed ZooKeeper, respectively, in order to 
provide reproducible and secure environments. The [Charmed Kafka snap](https://snapstore.io/charmed-kafka) and [Charmed ZooKeeper snap](https://snapstore.io/charmed-zookeeper) package the 
Kafka and ZooKeeper workload together with 
a set of dependencies and utilities required by the lifecycle of the operators (see [Charmed Kafka snap contents](https://github.com/canonical/charmed-kafka-snap/blob/3/edge/snap/snapcraft.yaml) and [Charmed ZooKeeper snap contents](https://github.com/canonical/charmed-zookeeper-snap/blob/3/edge/snap/snapcraft.yaml)). . 

> Every artifact bundled into the Charmed Kafka snap and Charmed ZooKeeper snap are verified against their MD5 SHA512 checksum after download. 

Author: yes, one of them. I somehow missed the "or"

# Cryptography

## Resource checksums

Every version of the Charmed Kafka and Charmed ZooKeeper operators install a pinned revision of the Charmed Kafka snap
and Charmed ZooKeeper, respectively, in order to 

> provide reproducible and secure environments. The Charmed Kafka snap and Charmed ZooKeeper snap package the 

Author: Done!

Kafka and ZooKeeper workload together with 

> a set of dependencies and utilities required by the lifecycle of the operators. 

Author: Done!

> Every artifact bundled into the Charmed Kafka snap and Charmed ZooKeeper snap are verified against their checksum. 

Author: I have add the specific of the checksum, but we do validate many artifacts that are part of the snap. I have also added the link of the code above, where people can actually see both checksum and the artifacts. 

PS. I tried to open the link but it seemed broken for me as the browser keeps hanging

## Sources verification

> Charmed Kafka and Charmed ZooKeeper artifacts are published programmatically using release pipelines implemented via GitHub Actions of the 

Author: I have revised this part quite a bit to make it more clear and also define better what is where (it was a comment from Rob as well)

associated repositories. 

> All repositories containing the source code of the Charmed Kafka project (i.e. charm and snap codebases) are set up with 

Author: Done!

> * Branch protection rules requiring additional commits to be merged via Pull-Request with at least 2 approvals from repository maintainers
> * Signed commits requirements (using GPG keys)
> * Requiring developers to sign CLA agreement

Reviewer: ```suggestion
* new commits to be merged via Pull-Request with at least 2 approvals from repository maintainers
* new commits to be signed using GPG keys
* developers to sign a Contributor License Agreement (CLA)
```

Can we provide a link here to one such agreement as an example?

Author: added the link

## Encryption

> The Charmed Kafka operator allow to deploy a secure Kafka cluster providing encryption-in-transit capabilities out of the box 

Reviewer: ```suggestion
The Charmed Kafka operator can be deployed as a secure Kafka cluster providing encryption-in-transit capabilities out of the box 
```

`allow` is in the [Style guide's](https://docs.ubuntu.com/styleguide/en#words-and-phrases-to-avoid) avoidance list. Have you tried the new content with Vale tests? There must be some false positives there, but all results are sorted by file so it should be easy enough to find errors and suggestions for this particular file. Alternatively, you can run it against a single file. But I never tried that. Maybe I should try it someday.

Author: done!

> for

Reviewer: ```suggestion
for:

```
The introductory sentence of a list should end with a colon if possible.

* Interbroker communications
* ZooKeeper connection
* External client connection 

In order to set up secure connection Kafka and ZooKeeper applications need to be integrated with TLS Certificate Provider charms, e.g. 
`self-signed-certificates` operator. CSRs are generated for every unit using `tls_certificates_interface` library that uses `cryptography` 
python library to create X.509 compatible certificates. The CSR is signed by the TLS Certificate Provider and returned to the units, and 
stored in a password-protected Keystore file. The password of the Keystore is stored in Juju secrets starting from revision 168 on Kafka 
and revision 130 on ZooKeeper. The relation provides also the certificate for the CA to be loaded in a password-protected Truststore file.

When encryption is enabled, hostname verification is turned on for client connections, including inter-broker communication. 

Encryption at rest is currently not supported, although it can be provided by the substrate (cloud or on-premises).

## Authentication

> There exists 3 layers of authentication in the Charmed Kafka solution:

Reviewer: ```suggestion
There are three layers of authentication in the Charmed Kafka solution:
```

Are we sure we want to call them layers? I'd expect layers to be enveloping each other, like in an onion. I'm not sure it is the case here, but maybe it is, that's why I'm asking.

Author: done! I have also restructured the text, althouhg using the word layer for authentication is not that weird, since auth is indeed a layer that is pluggable on top of endpoints (with different protocols)
